xonsh/ptk/key_bindings.py

```diff
@@ -41,3 +45,52 @@ def insert_literal_tab(event):
         """
         event.cli.current_buffer.insert_text(env.get('INDENT'))
 
+    @handle(Keys.ControlJ, filter=IsMultiline())
+    def multiline_carriage_return(event):
+        """
+        Preliminary parser to determine if 'Enter' key should send command to
+        the xonsh parser for execution or should insert a newline for continued
+        input.
+
+        Current 'triggers' for inserting a newline are:
+        - Not on first line of buffer and line is non-empty
+        - Previous character is a colon (covers if, for, etc...)
+        - User is in an open paren-block
+        - Line ends with backslash
+        - Any text exists below cursor position (relevant when editing previous
+        multiline blocks)
+        """
+
+        b = event.cli.current_buffer
+
+        # indent after a colon
+        if b.document.char_before_cursor == ':':
+            b.newline()
+            b.insert_text(indent_, fire_event=False)
+        # if current line isn't blank, check dedent tokens
+        elif (not (len(b.document.current_line) == 0 or
+                   b.document.current_line.isspace()) and
+              b.document.current_line.split(maxsplit=1)[0] in DEDENT_TOKENS):
+            b.newline(copy_margin=True)
+            _ = b.delete_before_cursor(count=len(indent_))
+        elif (not b.document.on_first_line and
+              not (len(b.document.current_line) == 0 or
+                   b.document.current_line.isspace())):
+            b.newline(copy_margin=True)
+        # if there are stray parens hanging aroung inside quotes
+        elif (re.search(parens, b.document.text) is not None and
+              ((b.document.text.count('(') -
+               re.search(parens, b.document.text).group().count('(')) >
+              (b.document.text.count(')') -
+               re.search(parens, b.document.text).group().count(')')))):
+            b.newline()
+        # and if there aren't stray parens hanging around inside quotes
+        elif (re.search(parens, b.document.text) is None and
+              b.document.text.count('(') > b.document.text.count(')')):
```

Comment:
Thanks for trying this @gforsyth. I still think that this is not general enough.  Really the best way is to just do a syntax check with the execer itself.  I have opened a PR to your fork, gforsyth/xonsh#1, that implements my vision for how this should work.  Merge it in if it looks good to you!